connor/antifraud/antifraud.go

```diff
@@ -0,0 +1,190 @@
+package antifraud
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/sonm-io/core/proto"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+type AntiFraud interface {
+	Run(ctx context.Context) error
+	DealOpened(deal *sonm.Deal) error
+	//DealClosed(deal *sonm.Deal)
+	TrackTask(ctx context.Context, deal *sonm.Deal, taskID string) error
+	//TaskDead(dealID *sonm.BigInt, taskID string)
+	FinishDeal(deal *sonm.Deal) error
+}
+
+type LogProcessor interface {
+	Run(ctx context.Context) error
+	TaskID() string
+	TaskQuality() (accurate bool, quality float64)
+}
+
+type dealMeta struct {
+	deal          *sonm.Deal
+	logProcessor  LogProcessor
+	poolProcessor LogProcessor
+}
+
+func lifeTime(deal *sonm.Deal) time.Duration {
+	return time.Now().Sub(deal.GetStartTime().Unix())
+}
+
+type antiFraud struct {
+	mu                sync.RWMutex
+	cfg               Config
+	meta              map[string]*dealMeta
+	blacklistWatchers map[common.Address]*blacklistWatcher
+	nodeConnection    *grpc.ClientConn
+	deals             sonm.DealManagementClient
+	log               *zap.Logger
+}
+
+func NewAntiFraud(cfg Config, log *zap.Logger, nodeConnection *grpc.ClientConn) AntiFraud {
+	return &antiFraud{
+		meta:              make(map[string]*dealMeta),
+		blacklistWatchers: map[common.Address]*blacklistWatcher{},
+		nodeConnection:    nodeConnection,
+		deals:             sonm.NewDealManagementClient(nodeConnection),
+		log:               log,
+		cfg:               cfg,
+	}
+}
+
+// Run blocks until context is cancelled or unrecoverable error met
+func (m *antiFraud) Run(ctx context.Context) error {
+	m.log.Info("starting antifraud")
+
+	ticker := time.NewTicker(m.cfg.QualityCheckInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			m.checkDeals(ctx)
+		}
+	}
+}
+
+//TODO: async
+func (m *antiFraud) checkDeals(ctx context.Context) error {
+	m.log.Debug("checking deals")
+	defer m.log.Debug("stop checking deals")
+
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	for _, dealMeta := range m.meta {
+		if dealMeta.logProcessor == nil {
+			m.log.Debug("skipping deal without task",
+				zap.String("deal_id", dealMeta.deal.GetId().Unwrap().String()))
+			continue
+		}
+
+		log := m.log.With(
+			zap.String("deal_id", dealMeta.deal.GetId().Unwrap().String()),
```

Comment:
Seems a little code duplicate with the same attribute assigning few lines upper.